reconstruction/rtpc/src/main/java/org/jlab/rec/rtpc/hit/KF/State.java

```diff
@@ -0,0 +1,2137 @@
+                        + alpha / kappa * (Math.cos(phi0) - Math.cos(phi0 + phi_minus));
+                double y = from.get_y() + drho * Math.sin(phi0)
+                        + alpha / kappa * (Math.sin(phi0) - Math.sin(phi0 + phi_minus));
+                double z = from.get_z() + dz - alpha / kappa * tanL * phi_minus;
+                // System.out.println( "r = " + Math.hypot(x,y) + " x = " + x + " y = " + y  + " z = " + z);
+
+                if (pl_step < min_pathLength)
+                    break;
+
+                kappa = energyLossb(kappa, tanL, pl_step, material);
+                phi_start = phi_minus;
+            }
+
+            pl += pl_s;
+
+        }
+
+        double alpha1 = to.alpha();
+
+        double Xc = x0 + (drho + alpha1 / kappa) * Math.cos(phi0);
+        double Yc = y0 + (drho + alpha1 / kappa) * Math.sin(phi0);
+        double phi0_prim = mod2PI(Math.atan2((Yc - y0_prim), (Xc - x0_prim)) - Math.PI);
+        double drho_prim = (Xc - x0_prim) * Math.cos(phi0_prim) + (Yc - y0_prim) * Math.sin(phi0_prim) - alpha1 / kappa;
+        double dz_prim = z0 - z0_prim + dz - (alpha1 / kappa) * (phi0_prim - phi0) * tanL;
+
+        Function func1 = (phi, kappa1) -> -end + Math.sqrt(Math
+                .pow(alpha1 * (Math.sin(phi0_prim) - Math.sin(phi + phi0_prim)) / kappa1 + drho_prim * Math.sin(phi0_prim) + y0_prim, 2)
+                + Math.pow(alpha1 * (Math.cos(phi0_prim) - Math.cos(phi + phi0_prim)) / kappa1 + drho_prim * Math.cos(phi0_prim) + x0_prim, 2));
+        double phi = minimize(bracket(0, 1, func1, kappa), func1, kappa);
+
+
+        return new State(drho_prim, phi0_prim, kappa, dz_prim, tanL, phi, pl);
+
+    }
+
+    public SimpleMatrix F(Site from, Site to, Material material) throws Exception {
+
+        double h = Math.sqrt(Math.ulp(1.0));
+        double[] dkappadx = new double[5];
+        SimpleMatrix sv = this._sv.copy();
+        double phi = this._phi;
+
+        for (int i = 0; i < 5; i++) {
+            SimpleMatrix sv_p = new SimpleMatrix(sv);
+            sv_p.set(i, 0, sv_p.get(i, 0) + h);
+            State state_p = new State(sv_p, phi);
+            State state_p_m = state_p.moveTo(from, to, material);
+            SimpleMatrix sv_p_m = state_p_m._sv.copy();
+
+            SimpleMatrix sv_m = new SimpleMatrix(sv);
+            sv_m.set(i, 0, sv_m.get(i, 0) - h);
+            State state_m = new State(sv_m, phi);
+            State state_m_m = state_m.moveTo(from, to, material);
+            SimpleMatrix sv_m_m = state_m_m._sv.copy();
+
+            double dkappadi = (sv_p_m.get(2, 0) - sv_m_m.get(2, 0)) / (2 * h);
+
+            dkappadx[i] = dkappadi;
+        }
+
+        double drho = this.drho();
+        double phi0 = this.phi0();
+        double kappa = this.kappa();
+        double tanL = this.tanL();
+
+        State a_prim = moveTo(from, to, material);
+
+        double alpha = from.alpha();
+
+        double drho_prim = a_prim.drho();
```

Comment:
Since you have different methods calculating the matrix F but I presume the equations defining the matrix elements are always the same, it may be better to implement a single method with the helix parameters of the initial and final states as inputs.

The same would apply to all calculations to get helix-related quantities like XC/YC, phi and so on, which should probably be methods of your Helix class